lib/mapObjects/JsonRandom.cpp

```diff
@@ -104,17 +125,52 @@ namespace JsonRandom
 		{
 			ret.push_back(loadValue(value[name], rng));
 		}
+		
+		if(!value["random"].isNull())
+		{
+			*RandomGeneratorUtil::nextItem(ret, rng) = loadValue(value["random"], rng);
+		}
+		
 		return ret;
 	}
 
 	std::map<SecondarySkill, si32> loadSecondary(const JsonNode & value, CRandomGenerator & rng)
 	{
 		std::map<SecondarySkill, si32> ret;
+		std::vector<SecondarySkill> except;
+		
 		for(const auto & pair : value.Struct())
 		{
-			SecondarySkill id(VLC->modh->identifiers.getIdentifier(pair.second.meta, "skill", pair.first).get());
-			ret[id] = loadValue(pair.second, rng);
+			std::string skillName = pair.first;
```

Comment:
Perhaps both primary and secondary can be implemented similar to 'loadResources' / 'loadKey' method, which would allow format like:
```js
[
    {    // gives 1 level of Wisdom skill...
        "type" : "wisdom", 
        "value" : 1 
    },
    {    // ... and 1-2 levels of a randomly selected magic school
        "list" : [ "airMagic", "waterMagic", "earthMagic", "fireMagic" ],
        "min" : 1, "max" 2
    }
]
```
I think we should eventually switch to vectors for all reward types where it makes sense.
And we should also account for skills that are potentially banned on the map.

Perhaps:
- if json uses struct / object form then load as before (and maybe consider than form deprecated)
- if json uses vector / list then for each entry in vector one skill will be generated:
- if entry contains 'type' then this skill will be selected
- else, if entry contains 'list' then one of these skills will be selected randomly (taking into account banned skills?)
- else, any non-banned random skill will be selected
- also, if entry contains 'except' entry then these skills should be except from selection

Reply:
@IvanSavenko but why vectors? Maybe struct of stucts to allow override in mods?

Reply:
Well, these vectors are already inside of another, much larger vector, so not much point. And replacing it is not trivial - we have order-dependent logic with it.

In long term - might make sense to replace. But don't see high priority. For now mods will have to replace entire rewards list.

Reply:
@IvanSavenko , I like this concept, the only headache I have - how to bypass map restrictions.
I'm thinking on adding global pointer to active map, is there better ideas?

Reply:
On naming - probably better to rename for consistency with other options which use similar naming:
- "list" -> "anyOf" ("oneOf" is also an option, but anyOf probably sounds better)
- "except" -> "noneOf"

Map restrictions - probably RewardableConstructor should pass reference to appropriate `std::vector<bool>` from CMap to JsonRandom functions, without passing entire CMap. Don't see any better way right now. 

There is (almost) suitable method that JsonRandom could have used - getDefaultAllowed() from handlers, but that method returns global settings, not map-specific.